File: python/src/x402_openai/server/providers.py
from __future__ import annotations

import json
from collections.abc import Mapping
from typing import Any, Literal, TypeAlias

from pydantic import BaseModel, ConfigDict

from x402_openai.core import TokenUsage, X402OpenAIError

ProviderApi: TypeAlias = Literal["responses", "chat"]


class ProviderSpec(BaseModel):
    model_config = ConfigDict(frozen=True)

    base_url: str
    api: ProviderApi
    key_env: str


# Major providers in the space. OpenAI is billed through its Responses API; every
# other provider here speaks the OpenAI-compatible Chat Completions API. Base URLs
# can be overridden per provider with <ID>_BASE_URL (e.g. GROQ_BASE_URL).
KNOWN_PROVIDERS: dict[str, ProviderSpec] = {
    "openai": ProviderSpec(
        base_url="https://api.openai.com/v1", api="responses", key_env="OPENAI_API_KEY"
    ),
    "openrouter": ProviderSpec(
        base_url="https://openrouter.ai/api/v1", api="chat", key_env="OPENROUTER_API_KEY"
    ),
    "gemini": ProviderSpec(
        base_url="https://generativelanguage.googleapis.com/v1beta/openai/",
        api="chat",
        key_env="GEMINI_API_KEY",
    ),
    "groq": ProviderSpec(
        base_url="https://api.groq.com/openai/v1", api="chat", key_env="GROQ_API_KEY"
    ),
    "deepseek": ProviderSpec(
        base_url="https://api.deepseek.com/v1", api="chat", key_env="DEEPSEEK_API_KEY"
    ),
    "xai": ProviderSpec(base_url="https://api.x.ai/v1", api="chat", key_env="XAI_API_KEY"),
    "mistral": ProviderSpec(
        base_url="https://api.mistral.ai/v1", api="chat", key_env="MISTRAL_API_KEY"
    ),
    "together": ProviderSpec(
        base_url="https://api.together.xyz/v1", api="chat", key_env="TOGETHER_API_KEY"
    ),
    "fireworks": ProviderSpec(
        base_url="https://api.fireworks.ai/inference/v1", api="chat", key_env="FIREWORKS_API_KEY"
    ),
    "perplexity": ProviderSpec(
        base_url="https://api.perplexity.ai", api="chat", key_env="PERPLEXITY_API_KEY"
    ),
    "cerebras": ProviderSpec(
        base_url="https://api.cerebras.ai/v1", api="chat", key_env="CEREBRAS_API_KEY"
    ),
}


class ProviderConfig(BaseModel):
    model_config = ConfigDict(frozen=True)

    id: str
    base_url: str
    api: ProviderApi
    api_key: str


def providers_from_env(env: Mapping[str, str]) -> tuple[ProviderConfig, ...]:
    """Build the enabled provider set from environment variables.

    A known provider is enabled when its API key variable is present. Custom or
    self-hosted OpenAI-compatible providers can be added through PROVIDERS_JSON,
    a JSON array of {"id", "base_url", "api", "api_key"} objects.
    """
    providers: list[ProviderConfig] = []
    for provider_id, spec in KNOWN_PROVIDERS.items():
        api_key = env.get(spec.key_env)
        if not api_key:
            continue
        base_url = env.get(f"{provider_id.upper()}_BASE_URL", spec.base_url)
        providers.append(
            ProviderConfig(id=provider_id, base_url=base_url, api=spec.api, api_key=api_key)
        )

    raw = env.get("PROVIDERS_JSON", "").strip()
    if raw:
        try:
            entries = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise ValueError("PROVIDERS_JSON must be valid JSON") from exc
        if not isinstance(entries, list):
            raise ValueError("PROVIDERS_JSON must be a JSON array")
        for entry in entries:
            if isinstance(entry, dict):
                # Accept snake_case (shared format) and camelCase (TS style) keys.
                entry = {
                    "id": entry.get("id"),
                    "base_url": entry.get("base_url", entry.get("baseUrl")),
                    "api": entry.get("api"),
                    "api_key": entry.get("api_key", entry.get("apiKey")),
                }
            providers.append(ProviderConfig.model_validate(entry))

    return tuple(providers)


def token_cap_error(api: ProviderApi, body: dict[str, Any], cap: int) -> str | None:
    """Validate the per-call output token limit for the provider's API flavor."""
    if api == "responses":
        field, value = "max_output_tokens", body.get("max_output_tokens")
    else:
        field = "max_tokens"
        value = body.get("max_tokens")
        if value is None:
            value = body.get("max_completion_tokens")
    if not isinstance(value, int) or isinstance(value, bool) or value <= 0:
        return f"{field} must be a positive integer"
    if value > cap:
        return f"{field} cannot exceed {cap}"
    return None


def extract_usage(api: ProviderApi, usage: dict[str, Any]) -> TokenUsage:
    """Normalize Responses and Chat Completions usage into billable token counts."""
    if api == "responses":
        details = usage.get("input_tokens_details") or {}
        return TokenUsage(
            input_tokens=usage.get("input_tokens", 0),
            output_tokens=usage.get("output_tokens", 0),
            cached_input_tokens=details.get("cached_tokens", 0) or 0,
        )
    details = usage.get("prompt_tokens_details") or {}
    return TokenUsage(
        input_tokens=usage.get("prompt_tokens", 0),
        output_tokens=usage.get("completion_tokens", 0),
        cached_input_tokens=details.get("cached_tokens", 0) or 0,
    )


async def create_upstream(
    client: Any,
    api: ProviderApi,
    body: dict[str, Any],
    request_id: str,
) -> tuple[dict[str, Any], TokenUsage, str | None]:
    """Call the upstream provider and return (response body, usage, request id)."""
    headers = {"X-Client-Request-Id": request_id}
    if api == "responses":
        result = await client.responses.create(**body, extra_headers=headers)
    else:
        result = await client.chat.completions.create(**body, extra_headers=headers)

    usage_object = getattr(result, "usage", None)
    if usage_object is None:
        raise X402OpenAIError("UPSTREAM_ERROR", "Upstream response did not contain usage", 502)

    usage = extract_usage(api, usage_object.model_dump())
    data = result.model_dump(mode="json", by_alias=True, exclude_none=True)
    return data, usage, getattr(result, "_request_id", None)

File: typescript/packages/server/src/providers.ts
import type OpenAI from "openai";
import type { ChatCompletionCreateParamsNonStreaming } from "openai/resources/chat/completions";
import type { ResponseCreateParamsNonStreaming } from "openai/resources/responses/responses";
import { z } from "zod";

import { X402OpenAIError } from "@infer402/core";
import type { TokenUsage } from "@infer402/core";

export type ProviderApi = "responses" | "chat";

export interface ProviderSpec {
  readonly baseUrl: string;
  readonly api: ProviderApi;
  readonly keyEnv: string;
}

// Major providers in the space. OpenAI is billed through its Responses API; every
// other provider here speaks the OpenAI-compatible Chat Completions API. Base URLs
// can be overridden per provider with <ID>_BASE_URL (e.g. GROQ_BASE_URL).
export const KNOWN_PROVIDERS = {
  openai: { baseUrl: "https://api.openai.com/v1", api: "responses", keyEnv: "OPENAI_API_KEY" },
  openrouter: { baseUrl: "https://openrouter.ai/api/v1", api: "chat", keyEnv: "OPENROUTER_API_KEY" },
  gemini: {
    baseUrl: "https://generativelanguage.googleapis.com/v1beta/openai/",
    api: "chat",
    keyEnv: "GEMINI_API_KEY",
  },
  groq: { baseUrl: "https://api.groq.com/openai/v1", api: "chat", keyEnv: "GROQ_API_KEY" },
  deepseek: { baseUrl: "https://api.deepseek.com/v1", api: "chat", keyEnv: "DEEPSEEK_API_KEY" },
  xai: { baseUrl: "https://api.x.ai/v1", api: "chat", keyEnv: "XAI_API_KEY" },
  mistral: { baseUrl: "https://api.mistral.ai/v1", api: "chat", keyEnv: "MISTRAL_API_KEY" },
  together: { baseUrl: "https://api.together.xyz/v1", api: "chat", keyEnv: "TOGETHER_API_KEY" },
  fireworks: {
    baseUrl: "https://api.fireworks.ai/inference/v1",
    api: "chat",
    keyEnv: "FIREWORKS_API_KEY",
  },
  perplexity: { baseUrl: "https://api.perplexity.ai", api: "chat", keyEnv: "PERPLEXITY_API_KEY" },
  cerebras: { baseUrl: "https://api.cerebras.ai/v1", api: "chat", keyEnv: "CEREBRAS_API_KEY" },
} satisfies Record<string, ProviderSpec>;

export const providerConfigSchema = z.object({
  id: z.string().min(1),
  baseUrl: z.string().url(),
  api: z.enum(["responses", "chat"]),
  apiKey: z.string().min(1),
});

export type ProviderConfig = z.infer<typeof providerConfigSchema>;

export function providersFromEnv(env: NodeJS.ProcessEnv): ProviderConfig[] {
  const providers: ProviderConfig[] = [];
  for (const [id, spec] of Object.entries(KNOWN_PROVIDERS)) {
    const apiKey = env[spec.keyEnv];
    if (!apiKey) continue;
    const baseUrl = env[`${id.toUpperCase()}_BASE_URL`] ?? spec.baseUrl;
    providers.push({ id, baseUrl, api: spec.api, apiKey });
  }

  const raw = (env.PROVIDERS_JSON ?? "").trim();
  if (raw) {
    let parsed: unknown;
    try {
      parsed = JSON.parse(raw);
    } catch {
      throw new X402OpenAIError("CONFIGURATION_ERROR", "PROVIDERS_JSON must be valid JSON");
    }
    if (!Array.isArray(parsed)) {
      throw new X402OpenAIError("CONFIGURATION_ERROR", "PROVIDERS_JSON must be a JSON array");
    }
    // Accept snake_case (shared with the Python gateway) and camelCase keys.
    const normalized = parsed.map((entry) => {
      const item = entry as Record<string, unknown>;
      return {
        id: item.id,
        baseUrl: item.baseUrl ?? item.base_url,
        api: item.api,
        apiKey: item.apiKey ?? item.api_key,
      };
    });
    providers.push(...z.array(providerConfigSchema).parse(normalized));
  }

  return providers;
}

export function tokenCapError(
  api: ProviderApi,
  body: Record<string, unknown>,
  cap: number,
): string | undefined {
  const field = api === "responses" ? "max_output_tokens" : "max_tokens";
  const value =
    api === "responses" ? body.max_output_tokens : (body.max_tokens ?? body.max_completion_tokens);
  if (typeof value !== "number" || !Number.isInteger(value) || value <= 0) {
    return `${field} must be a positive integer`;
  }
  if (value > cap) return `${field} cannot exceed ${cap}`;
  return undefined;
}

function toCount(value: unknown): number {
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

export function normalizeUsage(api: ProviderApi, usage: unknown): TokenUsage {
  const raw = (usage ?? {}) as Record<string, any>;
  if (api === "responses") {
    return {
      input_tokens: toCount(raw.input_tokens),
      output_tokens: toCount(raw.output_tokens),
      input_tokens_details: { cached_tokens: toCount(raw.input_tokens_details?.cached_tokens) },
    };
  }
  return {
    input_tokens: toCount(raw.prompt_tokens),
    output_tokens: toCount(raw.completion_tokens),
    input_tokens_details: { cached_tokens: toCount(raw.prompt_tokens_details?.cached_tokens) },
  };
}

export interface UpstreamResult {
  readonly data: unknown;
  readonly usage: TokenUsage;
  readonly requestId: string | undefined;
}

export async function createUpstream(
  client: OpenAI,
  api: ProviderApi,
  body: Record<string, unknown>,
  requestId: string,
): Promise<UpstreamResult> {
  const options = { headers: { "X-Client-Request-Id": requestId } };

  if (api === "responses") {
    const result = await client.responses
      .create(body as unknown as ResponseCreateParamsNonStreaming, options)
      .withResponse();
    if (!result.data.usage) {
      throw new X402OpenAIError("UPSTREAM_ERROR", "Upstream response did not contain usage", 502);
    }
    return {
      data: result.data,
      usage: normalizeUsage("responses", result.data.usage),
      requestId: result.response.headers.get("x-request-id") ?? undefined,
    };
  }

  const result = await client.chat.completions
    .create(body as unknown as ChatCompletionCreateParamsNonStreaming, options)
    .withResponse();
  if (!result.data.usage) {
    throw new X402OpenAIError("UPSTREAM_ERROR", "Upstream response did not contain usage", 502);
  }
  return {
    data: result.data,
    usage: normalizeUsage("chat", result.data.usage),
    requestId: result.response.headers.get("x-request-id") ?? undefined,
  };
}
